src/vs/workbench/api/browser/mainThreadTerminalService.ts

```diff
@@ -353,6 +354,13 @@ export class MainThreadTerminalService implements MainThreadTerminalServiceShape
 		this._proxy.$acceptTerminalTitleChange(terminalId, name);
 	}
 
+	private _onShellTypeChanged(terminalId: number, shellType: string): void {
+		const terminalInstance = this._terminalService.getInstanceFromId(terminalId)?.shellType;
+		if (terminalInstance) {
+			this._proxy.$acceptTerminalShellType(terminalId, shellType);
+		}
```

Comment:
You'll also want to set it when `shellType` is `undefined` or `''`, otherwise the value could get stale. For example moving from `bash` to an unknown shell

Reply:
Should be resolved with: https://github.com/microsoft/vscode/pull/237624/commits/53993cf02b6189f3d2e5117656ea9ea1d890c68e 

I intend to check if terminalInstance is not undefined and then set the shell type as whichever.

Reply:
This is because the event on terminal instance has data `TerminalShellType`. So we do need to map it in order to get the instance. What you have is good.

The alternative is what's done for `onData` which must include the data event as it's not stored on the instance:

https://github.com/microsoft/vscode/blob/8e584dec1f54713a9266a219cfa3b66031da359e/src/vs/workbench/contrib/terminal/browser/terminalService.ts#L160